compiler/rustc_borrowck/src/renumber.rs

```diff
@@ -77,7 +78,32 @@ impl<'a, 'tcx> MutVisitor<'tcx> for NllVisitor<'a, 'tcx> {
         debug!(?region);
     }
 
-    fn visit_const(&mut self, constant: &mut ty::Const<'tcx>, _location: Location) {
-        *constant = self.renumber_regions(*constant);
+    #[instrument(skip(self), level = "debug")]
+    fn visit_constant(&mut self, constant: &mut Constant<'tcx>, _location: Location) {
+        let literal = constant.literal;
```

Comment:
doesn't constant also implement `TypeFoldable`? so you can do `self.renumber_regions(*constant)`?

Reply:
If I use `constant.literal = self.renumber_regions(constant)` I get this ICE:

```
error: internal compiler error: /Users/bn/Documents/rust-local-fork/valtrees-cleanup/compiler/rustc_middle/src/ty/fold.rs:272:9: most type folders should not be folding MIR datastructures: Val(Scalar(<ZST>), for<'r> extern "rust-call" fn(&'r mut impl FnMut(A, B) -> T, (A, B)) -> <impl FnMut(A, B) -> T as ops::function::FnOnce<(A, B)>>::Output {<impl FnMut(A, B) -> T as ops::function::FnMut<(A, B)>>::call_mut})
```

Can't explain why that is.

Reply:
ah, because we still have the `fold_mir_constant`... call `constant.super_fold_with` for now I guess :sweat_smile:

Reply:
I do want to remove `fn fold_mir_constant` as well, because the only time that should be used is when we want to evaluate mir constants which should always happen explicitly

Reply:
> ah, because we still have the `fold_mir_constant`... call `constant.super_fold_with` for now I guess sweat_smile

this still needs to be done

Reply:
What folder do you want to fold the constant with?

Reply:
so the `NllVisitor` uses `infcx.tcx.fold_regions` instead of a custom folder.

I guess add `tcx.super_fold_regions` in the same file which uses `value.super_fold_with` instead? :thinking: